--- PackageKit-1.4.0/src/pk-dbus.c.orig	2026-09-09 22:42:50.000000000 +0200
+++ PackageKit-1.4.0/src/pk-dbus.c	2026-09-25 21:31:40.599431638 +0200
@@ -26,6 +26,7 @@
 #include <gio/gio.h>
 
 #ifdef HAVE_SYSTEMD_SD_LOGIN_H
+#include <systemd/sd-daemon.h>
 #include <systemd/sd-login.h>
 #endif
 
@@ -235,9 +236,7 @@ gchar *
 pk_dbus_get_session (PkDbus *dbus, const gchar *sender)
 {
 	gchar *session = NULL;
-#ifndef HAVE_SYSTEMD_SD_LOGIN_H
 	g_autoptr(GError) error = NULL;
-#endif
 	guint pid;
 	g_autoptr(GVariant) value = NULL;
 
@@ -260,10 +259,12 @@ pk_dbus_get_session (PkDbus *dbus, const
 
 	/* get session from systemd or ConsoleKit */
 #ifdef HAVE_SYSTEMD_SD_LOGIN_H
+	if (sd_booted () > 0) {
 	session = pk_dbus_get_session_systemd (pid);
 	if (session == NULL)
 		g_warning ("failed to get session for pid %u", pid);
-#else
+	} else {
+#endif
 	/* no ConsoleKit? */
 	if (dbus->proxy_session == NULL) {
 		g_warning ("no ConsoleKit, so cannot get session");
@@ -282,6 +283,8 @@ pk_dbus_get_session (PkDbus *dbus, const
 		goto out;
 	}
 	g_variant_get (value, "(o)", &session);
+#ifdef HAVE_SYSTEMD_SD_LOGIN_H
+	}
 #endif
 out:
 	return session;
